src/libcore/sync/atomic.rs

```diff
@@ -1101,7 +1104,7 @@ macro_rules! atomic_int {
         ///
         /// [module-level documentation]: index.html
         #[$stable]
-        #[repr(transparent)]
+        #[repr(align($align))]
```

Comment:
As above, I think that this is already always true with the exception of `u/i128`. I'd prefer a solution which kept `repr(transparent)` intact for other primitives and special-cased the 128-bit types to receive higher-alignment attributes rather than `#[repr(transparent)]`.

Reply:
`u64` is only aligned to 4 on 32-bit linux systems.

Reply:
@retep998 Thanks! I think I'd still prefer to special-case and allow `repr(transparent)` on more platforms, but that makes me unsure.

Reply:
Yeah, that's the issue, it's not clear which types can be `repr(transparent)`. That's why I just added `repr(align)` to all types to avoid the issue. We presumably definitely don't want the same type to have `repr(transparent)` on some platforms but not others so I don't really know what else to do here.

Reply:
I'm not sure either. Ping @rust-lang/libs for your thoughts? This is a breaking (but necessary) change to the representation guarantees of several atomics-- it'd be nice to keep the `repr(transparent)` where possible, but I agree that having it be platform-dependent is pretty dodgy.

Reply:
I'd personally be in favor of removing `#[repr(transparent)]` on the atomic types. I don't recall the motivation for adding them, though, so removing them to me is basically the more conservative thing to do (which is good for maintenance!)

Reply:
I did recently come across a need for atomic loads at specific addresses, which `#[repr(transparent)]` is a way to enable (if not the best way). Specifically the Linux perf_events interface uses a memory mapped ring buffer to transfer sample information between the kernel and user. The head and tail offsets corresponding to the "live" parts of the buffer are at specific offsets in the buffer, and since the kernel can be concurrently adjusting them you'd want to perform atomic loads. We could stabilize `std::ptr::atomic_load(*const u64, Ordering) -> u64` but without that the other option is to construct a `&AtomicU64` at the right address. This only works if `AtomicU64` has the same layout as u64, though.

Reply:
@sfackler I think though that the purpose here is that `atomic_load(u64)` for Linux only works on x86_64, not on x86 where it may not be aligned, right?

Reply:
Right, the alignment requirements are higher so `#[repr(transparent)]` may not be the best way to get to the functionality I'm looking for, but there is *some* use to there being representational guarantees about these types in the absence of other APIs.

Reply:
@sfackler I'd think you can still construct a `*const AtomicU64` at the right address without `repr(transparent)`?

Reply:
As long as the layout is the same, then `*const AtomicU64` to arbitrary memory is still fine. Where `repr(transparent)` is needed is specifically for compatibility in function calling convention ABIs because two types having the same layout may not be passed the same.

Reply:
My understanding is that `repr(transparent)` is the only thing that guarantees that the layout *is* the same - do we have a separate guarantee that single field structs have the same layout as their element?

Reply:
When not talking about passing things around in functions by value, wouldn't that be `repr(C)`?

Reply:
Yes, `repr(C)` would be one way to achieve layout compatibility. However, since this is in libstd, we can also just *state* that it's layout compatible and rely on the fact that rustc does newtype unwrapping. Furthermore, I am convinced we should guarantee layout compatibility of `repr(Rust)` newtypes anyway.